homeassistant/components/todoist/todo.py

```diff
@@ -127,6 +127,12 @@ async def async_update_todo_item(self, item: TodoItem) -> None:
         """Update a To-do item."""
         uid: str = cast(str, item.uid)
         if update_data := _task_api_data(item):
+            # In order to not lose any recurrence metadata for the task, we need to
+            # ensure that we send the `due_string` param if the task has it set.
+            api_data = next((d for d in self.coordinator.data if d.id == uid), None)
+            if api_data is not None:
+                if due := api_data.due:
+                    update_data["due_string"] = due.string
```

Comment:
Should this only be set when due_string was already present?

Reply:
If you set a due date/datetime this will always be set in my testing.

No due date:

```json
{
  "due": null
}
```

Due date:
```json
{
  "due": {
        "date": "2024-01-17",
        "is_recurring": false,
        "lang": "en",
        "string": "Jan 17"
    }
}
```

Due date/time:
```json
{
  "due": {
        "date": "2024-01-17",
        "datetime": "2024-01-17T21:30:00",
        "is_recurring": false,
        "lang": "en",
        "string": "Jan 17 9:30 PM"

    }
}
```
Due date recurring:

```json
{
  "due": {
        "date": "2024-01-17",
        "is_recurring": true,
        "lang": "en",
        "string": "every day"

    }
}
```

Reply:
I see, makes sense.

What should happen to the due string when due date is being changed?

Reply:
Great question.  I decided to move the logic inside of `_task_api_data` to avoid the situation where previously it would overwrite the `due_string` if the user had tried to unset it.

From testing it appears the `due_date` and `due_datetime` parameters take precedence over `due_string` if they don't match in the case where there is no recurrence.  For example this data POSTed to the endpoint when changing the due date from 2024-01-18 to 2024-01-31:

```json
{
  "content": "test",
  "description": "",
  "due_date": "2024-01-31",
  "due_string": "Jan 18"
}
```

And the task is updated correctly in todoist with the new due date of 2024-01-31.